### PR TITLE
Refactor the DSL to make it simpler to document

### PR DESCRIPTION
**Pull request by orta**
Two major changes

- Folder restructuring ( yawn ) 
- Separating the concerns of Danger's work, and Danger's DSL.

I was having issues thinking about how we could elegantly extend and document the DSL. We had these objects whose job it was to do work ( e.g. the `git` class does `exec` ) but to also provide DSL metadata e.g. (`commits`). That was compounded by @segiddins' very reasonable confusion that exposing commits as a public function would expose it to the DSL in #179, and ours too, we missed it on the PR.

So I've taken out the DSL functions into a separate object, whose job is to _just_ be the DSL. It acts similar to how our [ViewModels](https://github.com/artsy/eigen/blob/master/Artsy/View_Controllers/Live_Auctions/ViewModels/LiveAuctionLotViewModel.swift#L87) work in Eigen. Offering a nice separation of concerns. 

Doing that meant I could build on some work I did for [gh-issues-inspector](https://github.com/orta/gh-issues-inspector) which used yard to [generate the README](https://github.com/orta/gh-issues-inspector/blob/master/Rakefile#L18) from inline code comments.

So I have a rough example version that does this:

``` sh
~/d/r/l/danger (crash_bang) ⏛  rake doc
## Danger

#### Core
- markdown
  : Print markdown to below the table
  : lib/danger/danger_core/dangerfile_dsl.rb - 15

- message
  : Print out a generate message on the PR
  : lib/danger/danger_core/dangerfile_dsl.rb - 27

- warn
  : Specifies a problem, but not critical
  : lib/danger/danger_core/dangerfile_dsl.rb - 39

- fail
  : Declares a CI blocking error
  : lib/danger/danger_core/dangerfile_dsl.rb - 53

#### Plugins
- import
  : Download a local or remote plugin and use it locally
  : lib/danger/danger_core/dangerfile_dsl.rb - 65

- import_url
  : Download a remote plugin and use it locally
  : lib/danger/danger_core/dangerfile_dsl.rb - 81

- import_local
  : Import one or more local plugins
  : lib/danger/danger_core/dangerfile_dsl.rb - 107

#### Danger Zone
- env
  : Provides access to the raw Travis/Circle/Buildkite/GitHub objects, which
you can use to pull out extra bits of information. _Warning_
the interfaces of these objects is **not** considered a part of the Dangerfile public
API, and is viable to change occasionally on the whims of developers.
  : lib/danger/danger_core/dangerfile_dsl.rb - 120

#### Git Files
- added_files
  : Paths for files that were added during the diff
  : lib/danger/scm_source/git_repo.rb - 15

- deleted_files
  : Paths for files that were removed during the diff
  : lib/danger/scm_source/git_repo.rb - 23

- modified_files
  : Paths for files that changed during the diff
  : lib/danger/scm_source/git_repo.rb - 31

#### Git Metadata
- lines_of_code
  : The overall lines of code added/removed in the diff
  : lib/danger/scm_source/git_repo.rb - 39

- deletions
  : The overall lines of code removed in the diff
  : lib/danger/scm_source/git_repo.rb - 47

- insertions
  : The overall lines of code added in the diff
  : lib/danger/scm_source/git_repo.rb - 55

- commits
  : The log of commits inside the diff
  : lib/danger/scm_source/git_repo.rb - 63

#### PR Metadata
- pr_title
  : The title of the Pull Request
  : lib/danger/request_source/github.rb - 17

- pr_body
  : The body text of the Pull Request
  : lib/danger/request_source/github.rb - 25

- pr_author
  : The username of the author of the Pull Request
  : lib/danger/request_source/github.rb - 33

- pr_labels
  : The labels assigned to the Pull Request
  : lib/danger/request_source/github.rb - 41

#### PR Commit Metadata
- branch_for_merge
  : The branch to which the PR is going to be merged into
  : lib/danger/request_source/github.rb - 49

- base_commit
  : The base commit to which the PR is going to be merged as a parent
  : lib/danger/request_source/github.rb - 57

- head_commit
  : The head commit to which the PR is requesting to be merged from
  : lib/danger/request_source/github.rb - 65
```

As a current proof of concept, and I've expanded `rake spec` to also verify that all public DSL items are publicly documented too. ✈️

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -45,3 +45,4 @@ build/
 .rvmrc
 fastlane/report.xml
 .keys
+.idea
```
Comment (jeroenvisser101) [inline .gitignore]: You might want to add editor-specific ignores to your global gitignore instead

File: lib/danger/commands/runner.rb [hunk 2]
```diff
@@ -54,6 +54,19 @@ def run
...
 
+    def self.report_error(exception)
+      case exception
+      when Interrupt
+        puts '[!] Cancelled'.red
+        Config.instance.verbose? ? raise : exit(1)
+      when SystemExit
+        raise
+      else
+        # puts UI::ErrorReport.report(exception)
```
Comment (KrauseFx) [inline lib/danger/commands/runner.rb]: Temporary code?

Comment (orta) [inline lib/danger/commands/runner.rb]: Yeah, it was to add support for gh-issues-inspector, it should get removed and that can be a separate pr

File: .gitignore (hunk 1 above)
Comment (orta) [inline .gitignore]: done

File: lib/danger/commands/runner.rb (hunk 2 above)
Comment (KrauseFx) [inline lib/danger/commands/runner.rb]: :+1:

**Comment by orta** [comment]
☁️ 

Took 3 hours to get green.

💚

File: spec/commands/runner_command_spec.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        Dir.chdir dir do
+          `git init`
+          # `git remote add origin git@github.com:artsy/eigen.git`
+          `touch Dangerfile`
+          expect { Danger::Runner.run([]) }.to_not raise_error SystemExit
+        end
+      end
+    end
+  end
+end
```
Comment (orta) [inline spec/commands/runner_command_spec.rb]: added some initial `danger` test runs as a proof of concept 👍

**Pull request merged**